File: packages/python/onnx9000-onnx2gguf/src/onnx9000/onnx2gguf/builder.py
"""Module providing core GGUF Builder logic."""

import struct
from enum import Enum
from typing import Any, BinaryIO, Union


class GGUFValueType(Enum):
    """GGUFValueType implementation."""

    UINT8 = 0
    INT8 = 1
    UINT16 = 2
    INT16 = 3
    UINT32 = 4
    INT32 = 5
    FLOAT32 = 6
    BOOL = 7
    STRING = 8
    ARRAY = 9
    UINT64 = 10
    INT64 = 11
    FLOAT64 = 12


class GGUFTensorType(Enum):
    """GGUFTensorType implementation."""

    F32 = 0
    F16 = 1
    Q4_0 = 2
    Q4_1 = 3
    Q8_0 = 8


class GGUFWriter:
    """Zero-dependency GGUF Writer."""

    def __init__(self, f: BinaryIO):
        """Perform   init   operation."""
        self.f = f
        self.kvs: list[tuple[str, GGUFValueType, Any]] = []
        self.tensors: list[dict[str, Any]] = []
        self.tensor_data_offset = 0
        self.tensor_sizes = []

    def add_uint8(self, key: str, val: int) -> None:
        """Perform add uint8 operation."""
        self.kvs.append((key, GGUFValueType.UINT8, val))

    def add_int8(self, key: str, val: int) -> None:
        """Perform add int8 operation."""
        self.kvs.append((key, GGUFValueType.INT8, val))

    def add_uint16(self, key: str, val: int) -> None:
        """Perform add uint16 operation."""
        self.kvs.append((key, GGUFValueType.UINT16, val))

    def add_int16(self, key: str, val: int) -> None:
        """Perform add int16 operation."""
        self.kvs.append((key, GGUFValueType.INT16, val))

    def add_uint32(self, key: str, val: int) -> None:
        """Perform add uint32 operation."""
        self.kvs.append((key, GGUFValueType.UINT32, val))

    def add_int32(self, key: str, val: int) -> None:
        """Perform add int32 operation."""
        self.kvs.append((key, GGUFValueType.INT32, val))

    def add_float32(self, key: str, val: float) -> None:
        """Perform add float32 operation."""
        self.kvs.append((key, GGUFValueType.FLOAT32, val))

    def add_uint64(self, key: str, val: int) -> None:
        """Perform add uint64 operation."""
        self.kvs.append((key, GGUFValueType.UINT64, val))

    def add_int64(self, key: str, val: int) -> None:
        """Perform add int64 operation."""
        self.kvs.append((key, GGUFValueType.INT64, val))

    def add_float64(self, key: str, val: float) -> None:
        """Perform add float64 operation."""
        self.kvs.append((key, GGUFValueType.FLOAT64, val))

    def add_bool(self, key: str, val: bool) -> None:
        """Perform add bool operation."""
        self.kvs.append((key, GGUFValueType.BOOL, val))

    def add_string(self, key: str, val: str) -> None:
        """Perform add string operation."""
        self.kvs.append((key, GGUFValueType.STRING, val))

    def add_array(self, key: str, val: list[Any], array_type: GGUFValueType) -> None:
        """Perform add array operation."""
        self.kvs.append((key, GGUFValueType.ARRAY, (array_type, val)))

    def _write_string(self, s: str) -> None:
        """Perform  write string operation."""
        encoded = s.encode("utf-8")
        if len(encoded) > 2**31 - 1:
            raise ValueError("String exceeds standard allocation limits")
        self.f.write(struct.pack("<Q", len(encoded)))
        self.f.write(encoded)

    def _write_val(self, vtype: GGUFValueType, val: Any) -> None:
        """Perform  write val operation."""
        if vtype == GGUFValueType.UINT8:
            self.f.write(struct.pack("<B", val))
        elif vtype == GGUFValueType.INT8:
            self.f.write(struct.pack("<b", val))
        elif vtype == GGUFValueType.UINT16:
            self.f.write(struct.pack("<H", val))
        elif vtype == GGUFValueType.INT16:
            self.f.write(struct.pack("<h", val))
        elif vtype == GGUFValueType.UINT32:
            self.f.write(struct.pack("<I", val))
        elif vtype == GGUFValueType.INT32:
            self.f.write(struct.pack("<i", val))
        elif vtype == GGUFValueType.FLOAT32:
            self.f.write(struct.pack("<f", float(val)))
        elif vtype == GGUFValueType.UINT64:
            self.f.write(struct.pack("<Q", val))
        elif vtype == GGUFValueType.INT64:
            self.f.write(struct.pack("<q", val))
        elif vtype == GGUFValueType.FLOAT64:
            self.f.write(struct.pack("<d", float(val)))
        elif vtype == GGUFValueType.BOOL:
            self.f.write(struct.pack("<?", bool(val)))
        elif vtype == GGUFValueType.STRING:
            self._write_string(val)
        elif vtype == GGUFValueType.ARRAY:
            (atype, arr) = val
            self.f.write(struct.pack("<I", atype.value))
            self.f.write(struct.pack("<Q", len(arr)))
            for item in arr:
                self._write_val(atype, item)
        else:
            raise ValueError(f"Unknown value type: {vtype}")

    def add_tensor_info(self, name: str, shape: list[int], dtype: GGUFTensorType) -> None:
        """Perform add tensor info operation."""
        if not isinstance(dtype, GGUFTensorType):
            raise ValueError("Invalid tensor type")
        self.tensors.append(
            {"name": name, "shape": shape, "type": dtype.value, "offset": self.tensor_data_offset}
        )
        items = 1
        for d in shape:
            items *= d
        if dtype == GGUFTensorType.F32:
            size = items * 4
        elif dtype == GGUFTensorType.F16:
            size = items * 2
        elif dtype == GGUFTensorType.Q4_0:
            assert items % 32 == 0, "Q4_0 requires multiples of 32"
            size = items // 32 * 18
        elif dtype == GGUFTensorType.Q4_1:
            assert items % 32 == 0, "Q4_1 requires multiples of 32"
            size = items // 32 * 20
        else:
            assert items % 32 == 0, "Q8_0 requires multiples of 32"
            size = items // 32 * 34
        alignment = 32
        for k, vtype, val in self.kvs:
            if k == "general.alignment" and vtype == GGUFValueType.UINT32:
                alignment = val
        padding = (alignment - size % alignment) % alignment
        self.tensor_sizes.append(size + padding)
        self.tensor_data_offset += size + padding

    def write_header_to_file(self) -> None:
        """Perform write header to file operation."""
        self.f.write(b"GGUF")
        self.f.write(struct.pack("<I", 3))
        self.f.write(struct.pack("<Q", len(self.tensors)))
        self.f.write(struct.pack("<Q", len(self.kvs)))
        for k, vtype, val in self.kvs:
            self._write_string(k)
            self.f.write(struct.pack("<I", vtype.value))
            self._write_val(vtype, val)
        for t in self.tensors:
            self._write_string(t["name"])
            self.f.write(struct.pack("<I", len(t["shape"])))
            for dim in t["shape"]:
                self.f.write(struct.pack("<Q", dim))
            self.f.write(struct.pack("<I", t["type"]))
            self.f.write(struct.pack("<Q", t["offset"]))
        alignment = 32
        for k, vtype, val in self.kvs:
            if k == "general.alignment" and vtype == GGUFValueType.UINT32:
                alignment = val
        current_pos = self.f.tell()
        padding = (alignment - current_pos % alignment) % alignment
        self.f.write(b"\x00" * padding)

    def write_tensor_data(self, tensor_bytes: bytes) -> None:
        """Perform write tensor data operation."""
        self.f.write(tensor_bytes)
        alignment = 32
        for k, vtype, val in self.kvs:
            if k == "general.alignment" and vtype == GGUFValueType.UINT32:
                alignment = val
        padding = (alignment - len(tensor_bytes) % alignment) % alignment
        self.f.write(b"\x00" * padding)

File: packages/js/onnx2gguf/src/builder.ts
/* eslint-disable */
export enum GGUFValueType {
  UINT8 = 0,
  INT8 = 1,
  UINT16 = 2,
  INT16 = 3,
  UINT32 = 4,
  INT32 = 5,
  FLOAT32 = 6,
  BOOL = 7,
  STRING = 8,
  ARRAY = 9,
  UINT64 = 10,
  INT64 = 11,
  FLOAT64 = 12,
}

export enum GGUFTensorType {
  F32 = 0,
  F16 = 1,
  Q4_0 = 2,
  Q4_1 = 3,
  Q8_0 = 8,
}

export class GGUFWriter {
  public kvs: { key: string; type: GGUFValueType; val: ReturnType<typeof JSON.parse> }[] = [];
  public tensors: { name: string; shape: bigint[]; type: GGUFTensorType; offset: bigint }[] = [];

  public addUint8(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.UINT8, val });
  }
  public addInt8(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.INT8, val });
  }
  public addUint16(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.UINT16, val });
  }
  public addInt16(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.INT16, val });
  }
  public addUint32(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.UINT32, val });
  }
  public addInt32(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.INT32, val });
  }
  public addFloat32(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.FLOAT32, val });
  }
  public addUint64(key: string, val: bigint): void {
    this.kvs.push({ key, type: GGUFValueType.UINT64, val });
  }
  public addInt64(key: string, val: bigint): void {
    this.kvs.push({ key, type: GGUFValueType.INT64, val });
  }
  public addFloat64(key: string, val: number): void {
    this.kvs.push({ key, type: GGUFValueType.FLOAT64, val });
  }
  public addBool(key: string, val: boolean): void {
    this.kvs.push({ key, type: GGUFValueType.BOOL, val });
  }
  public addString(key: string, val: string): void {
    this.kvs.push({ key, type: GGUFValueType.STRING, val });
  }
  public addArray(
    key: string,
    val: ReturnType<typeof JSON.parse>[],
    arrayType: GGUFValueType,
  ): void {
    this.kvs.push({ key, type: GGUFValueType.ARRAY, val: { arrayType, items: val } });
  }

  public addTensorInfo(name: string, shape: bigint[], type: GGUFTensorType, offset: bigint): void {
    this.tensors.push({ name, shape, type, offset });
  }

  public getHeaderSize(): number {
    let size = 0;
    size += 4; // Magic
    size += 4; // Version
    size += 8; // Tensor count
    size += 8; // KV count

    // KVs
    const encoder = new TextEncoder();
    for (const kv of this.kvs) {
      size += 8; // key string len
      size += encoder.encode(kv.key).length;
      size += 4; // type
      size += this.getValueSize(kv.type, kv.val);
    }

    // Tensors
    for (const t of this.tensors) {
      size += 8; // name string len
      size += encoder.encode(t.name).length;
      size += 4; // n_dims
      size += t.shape.length * 8; // dims
      size += 4; // type
      size += 8; // offset
    }

    let alignment = 32;
    for (const kv of this.kvs) {
      if (kv.key === 'general.alignment' && kv.type === GGUFValueType.UINT32) {
        alignment = kv.val as number;
      }
    }
    const padding = (alignment - (size % alignment)) % alignment;
    size += padding;
    return size;
  }

  private getValueSize(type: GGUFValueType, val: ReturnType<typeof JSON.parse>): number {
    switch (type) {
      case GGUFValueType.UINT8:
      case GGUFValueType.INT8:
      case GGUFValueType.BOOL:
        return 1;
      case GGUFValueType.UINT16:
      case GGUFValueType.INT16:
        return 2;
      case GGUFValueType.UINT32:
      case GGUFValueType.INT32:
      case GGUFValueType.FLOAT32:
        return 4;
      case GGUFValueType.UINT64:
      case GGUFValueType.INT64:
      case GGUFValueType.FLOAT64:
        return 8;
      case GGUFValueType.STRING: {
        const encoder = new TextEncoder();
        return 8 + encoder.encode(val).length;
      }
      case GGUFValueType.ARRAY: {
        let s = 4; // array type
        s += 8; // array len
        const atype = val.arrayType as GGUFValueType;
        for (const item of val.items) {
          s += this.getValueSize(atype, item);
        }
        return s;
      }
      default:
        /* v8 ignore start */
        throw new Error('Unknown type');
      /* v8 ignore stop */
    }
  }

  public writeHeader(buffer: ArrayBuffer, byteOffset: number = 0): number {
    const view = new DataView(buffer, byteOffset);
    let offset = 0;
    const encoder = new TextEncoder();

    const writeString = (s: string) => {
      const encoded = encoder.encode(s);
      view.setBigUint64(offset, BigInt(encoded.length), true);
      offset += 8;
      const u8 = new Uint8Array(buffer, byteOffset + offset, encoded.length);
      u8.set(encoded);
      offset += encoded.length;
    };

    const writeVal = (vtype: GGUFValueType, val: ReturnType<typeof JSON.parse>) => {
      switch (vtype) {
        case GGUFValueType.UINT8:
          view.setUint8(offset, val);
          offset += 1;
          break;
        case GGUFValueType.INT8:
          view.setInt8(offset, val);
          offset += 1;
          break;
        case GGUFValueType.UINT16:
          view.setUint16(offset, val, true);
          offset += 2;
          break;
        case GGUFValueType.INT16:
          view.setInt16(offset, val, true);
          offset += 2;
          break;
        case GGUFValueType.UINT32:
          view.setUint32(offset, val, true);
          offset += 4;
          break;
        case GGUFValueType.INT32:
          view.setInt32(offset, val, true);
          offset += 4;
          break;
        case GGUFValueType.FLOAT32:
          view.setFloat32(offset, val, true);
          offset += 4;
          break;
        case GGUFValueType.UINT64:
          view.setBigUint64(offset, val, true);
          offset += 8;
          break;
        case GGUFValueType.INT64:
          view.setBigInt64(offset, val, true);
          offset += 8;
          break;
        case GGUFValueType.FLOAT64:
          view.setFloat64(offset, val, true);
          offset += 8;
          break;
        case GGUFValueType.BOOL:
          view.setUint8(offset, val ? 1 : 0);
          offset += 1;
          break;
        case GGUFValueType.STRING:
          writeString(val);
          break;
        case GGUFValueType.ARRAY: {
          const atype = val.arrayType as GGUFValueType;
          const items = val.items as ReturnType<typeof JSON.parse>[];
          view.setUint32(offset, atype, true);
          offset += 4;
          view.setBigUint64(offset, BigInt(items.length), true);
          offset += 8;
          for (const item of items) {
            writeVal(atype, item);
          }
          break;
        }
        default:
          /* v8 ignore start */
          throw new Error('Unknown type');
        /* v8 ignore stop */
      }
    };

    // Magic: GGUF
    view.setUint8(offset++, 0x47); // G
    view.setUint8(offset++, 0x47); // G
    view.setUint8(offset++, 0x55); // U
    view.setUint8(offset++, 0x46); // F

    // Version 3
    view.setUint32(offset, 3, true);
    offset += 4;

    // Counts
    view.setBigUint64(offset, BigInt(this.tensors.length), true);
    offset += 8;
    view.setBigUint64(offset, BigInt(this.kvs.length), true);
    offset += 8;

    // KVs
    for (const kv of this.kvs) {
      writeString(kv.key);
      view.setUint32(offset, kv.type, true);
      offset += 4;
      writeVal(kv.type, kv.val);
    }

    // Tensors
    for (const t of this.tensors) {
      writeString(t.name);
      view.setUint32(offset, t.shape.length, true);
      offset += 4;
      for (const dim of t.shape) {
        view.setBigUint64(offset, dim, true);
        offset += 8;
      }
      view.setUint32(offset, t.type, true);
      offset += 4;
      view.setBigUint64(offset, t.offset, true);
      offset += 8;
    }

    let alignment = 32;
    for (const kv of this.kvs) {
      if (kv.key === 'general.alignment' && kv.type === GGUFValueType.UINT32) {
        alignment = kv.val;
      }
    }

    const padding = (alignment - (offset % alignment)) % alignment;
    for (let i = 0; i < padding; i++) {
      view.setUint8(offset++, 0);
    }

    return offset;
  }
}
